### ISSUE TITLE
Clang doesn't generate a function when using __array_rank

### ISSUE THREAD

**Issue by philnik777** [llvm/llvm-project#57133]
The code [(Godbolt)](https://godbolt.org/z/3qe4Gqqvx)
```c++
#include <__type_traits/integral_constant.h>

template <class T>
struct rank : std::integral_constant<int, __array_rank(T)> {};

template <class T>
void test_rank() {
  static_assert(rank<T>::value == 0);
}

void func() {
  test_rank<void>();
}
```
generates `func()`, but doesn't generate `test_rank<void>()`.

**Comment by cjdb** [comment]
Potentially related: [return statement in `f` is ignored](https://compiler-explorer.com/z/rWxW1EMKP).

```cpp
template<auto v>
struct integral_constant {
    static inline constexpr auto value = v;
};

template<class T>
struct rank : integral_constant<__array_rank(T)> {};

__SIZE_TYPE__ f()
{
  return rank<int[1]>::value;
}
```

**Comment by philnik777** [comment]
Looks like this got fixed in 19: https://godbolt.org/z/WWqdnnYPs

**Comment by philnik777** [comment]
Looks like this wasn't actually fixed until 20, see #123498.
